### PR TITLE
Nonatomic payment processing

### PR DESCRIPTION
**Pull request by rlucioni**
Allows payment processing views to commit to the database at each step along the way to order placement. These views perform three major functions: 1) recording raw processor responses, 2) recording successful payments by creating payment events and sources, and 3) placing orders. If any of these operations fails, we don't want any prior operations to be rolled back.

Additionally, if a failure occurs at any point in this process, dirty data shouldn't be introduced to the system. Processor responses, payment events and sources, and orders can each be introduced independently, one after the other. We don't rely on the atomic creation of these data; in fact, it's exactly what we'd like to avoid.

This addresses [XCOM-415](https://openedx.atlassian.net/browse/XCOM-415). @jimabramson or @clintonb, please review.

### CODE REVIEW

File: ecommerce/extensions/payment/tests/test_views.py [hunk 1]
```diff
@@ -270,9 +295,17 @@ def test_payment_error(self):
         """
         with mock.patch.object(PaypalPaymentExecutionView, 'handle_payment',
                                side_effect=PaymentError) as fake_handle_payment:
-            self._assert_execution_redirect()
+            creation_response, _ = self._assert_execution_redirect()
```
Comment (clintonb) [inline ecommerce/extensions/payment/tests/test_views.py]: Use a double underscore to avoid potential issues with `ugettext`.

File: ecommerce/extensions/payment/views.py [hunk 2]
```diff
@@ -81,6 +83,12 @@ def post(self, request):
             order_number = cybersource_response.get('req_reference_number')
             basket_id = OrderNumberGenerator().basket_id(order_number)
 
+            logger.info(
+                'Received CyberSource merchant notification for basket [%d], to be associated with order [%s].',
```
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: This is redundant information. Choose one. Also, it is crucial that the CyberSource transaction ID be logged in the event of another failure to store this information in the database.

File: ecommerce/extensions/payment/tests/test_views.py (hunk 1 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/tests/test_views.py]: Updated.

**Comment by clintonb** [comment]
You've disabled atomicity for the views as a whole, but you have not restored atomicity inside the view functions (after the payment processor response is recorded).

File: ecommerce/extensions/payment/views.py [hunk 3]
```diff
@@ -29,6 +30,7 @@
 PaymentProcessorResponse = get_model('payment', 'PaymentProcessorResponse')
 
 
+@transaction.non_atomic_requests
```
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: This should decorate the post method.

Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: I agree based on a look at the [docs](https://docs.djangoproject.com/en/1.8/topics/db/transactions/#django.db.transaction.non_atomic_requests).  The fact that the class is decorated, but the tests still pass, is confusing to me.

Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: My mistake, I've corrected this. @jimabramson decorating the class appears to disable atomicity for every view method within, including the `post` method.

File: ecommerce/extensions/payment/views.py (hunk 2 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Thanks, good catch. I've updated this and added tests to cover it. The tests have exposed some holes in our logging, but patching those is out of scope for this work and should be addressed by [XCOM-381](https://openedx.atlassian.net/browse/XCOM-381).

**Comment by rlucioni** [comment]
@clintonb I am intentionally refraining from restoring atomicity within the (now correctly decorated) view functions for reasons described above in the description of this PR, and in the comment thread for [XCOM-415](https://openedx.atlassian.net/browse/XCOM-415). In a nutshell, we want these views to be able to commit to the database at each step along the way to order placement.

**Comment by jimabramson** [comment]
@rlucioni if the process of placing or fulfilling the order involves more than one database insert/update it must be done atomically.  Otherwise a failure in between those operations will leave a mess.

**Comment by clintonb** [comment]
What @jimabramson said. We don't want dirty data in the DB.

**Comment by rlucioni** [comment]
@jimabramson and @clintonb you're right; I was thinking of order placement and fulfillment as an atomic operation, when it is in fact not. I'll update to make sure those operations are handled atomically.

**Comment by rlucioni** [comment]
Partial atomicity has made its triumphant return, along with tests. Ready for re-review.

**Comment by clintonb** [comment]
:+1: (Awaiting second review from @jimabramson)

File: ecommerce/extensions/payment/views.py [hunk 4]
```diff
@@ -156,28 +172,34 @@ def get(self, request):
...
+                        billing_address=None,
+                        order_total=order_total
+                    )
+                except UnableToPlaceOrder:
+                    logger.exception('Payment was executed, but an order was not created for basket [%d].', basket.id)
+
+                return redirect(receipt_url)
+        except:  # pylint: disable=bare-except
+            logger.exception('Payment was received, but attempts to create an order for basket [%d] failed.', basket.id)
+            return redirect(receipt_url)
```
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py:218]: so when order creation fails, the user will get the checkout receipt page, which will time out after 5 attempts to fetch the order and draw a receipt (right)?    not saying this is better/worse than the present state, just confirming the expectation.

i doubt it's worth it, but we could consider sending the user to the receipt without a basket_id parameter.   this would at least have the effect of skipping the 5 attempts at retrieving the order.

File: ecommerce/extensions/payment/views.py [hunk 5]
```diff
@@ -103,27 +111,34 @@ def post(self, request):
...
+                except UnableToPlaceOrder:
+                    logger.exception('Payment was received, but an order was not created for basket [%d].', basket.id)
+                    # Ensure we return, in case future changes introduce post-order placement functionality.
+                    return HttpResponse()
 
-        return HttpResponse()
+                return HttpResponse()
+        except:  # pylint: disable=bare-except
+            logger.exception('Payment was received, but attempts to create an order for basket [%d] failed.', basket.id)
+            return HttpResponse()
```
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: is there any reason not to respond with status 500 here?   it's not user facing, but i think it would increase the likelihood of noticing problems quickly.

File: ecommerce/extensions/payment/views.py (hunk 4 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py:218]: @jimabramson yes, that's the expected behavior. I don't think the added complexity of sending the user to the receipt without a basket ID is worth it. One way or another, they're going to encounter an error message.

File: ecommerce/extensions/payment/views.py (hunk 5 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Good idea. Updated.

**Comment by jimabramson** [comment]
shouldn't the call to `handle_payment` (https://github.com/edx/ecommerce/blob/renzo/nonatomic-payment-processing/ecommerce/extensions/payment/views.py#L103) also get wrapped in `atomic`?   it's doing two inserts.

**Comment by rlucioni** [comment]
@jimabramson yes, nice catch. Sources and payment events shouldn't be created separately here. Updated.

File: ecommerce/extensions/payment/views.py [hunk 6]
```diff
@@ -92,38 +100,54 @@ def post(self, request):
...
+                billing_address = self._get_billing_address(cybersource_response)
+
+                try:
+                    user = basket.owner
+                    self.handle_order_placement(order_number, user, basket, None, shipping_method, shipping_charge,
+                                                billing_address, order_total)
+                except UnableToPlaceOrder:
+                    logger.exception('Payment was received, but an order was not created for basket [%d].', basket.id)
+                    # Ensure we return, in case future changes introduce post-order placement functionality.
+                    return HttpResponse()
```
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: sorry for missing this before, but I think this should be a 500 as well.

File: ecommerce/extensions/payment/views.py [hunk 7]
```diff
@@ -92,38 +100,54 @@ def post(self, request):
...
+                        ppr.id
+                    )
+                    return HttpResponse(status=400)
+                except PaymentError:
+                    logger.exception(
+                        'CyberSource payment failed for basket [%d]. The payment response was recorded in entry [%d].',
+                        basket.id,
+                        ppr.id
+                    )
+                    return HttpResponse()
```
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py:119]: same here, 500

Comment (jimabramson) [inline ecommerce/extensions/payment/views.py:119]: also, one more here https://github.com/edx/ecommerce/blob/renzo/nonatomic-payment-processing/ecommerce/extensions/payment/views.py#L96.

if you prefer to make those changes in another PR i'm ok with that

**Comment by jimabramson** [comment]
@rlucioni :+1: whether you decide to add those 500s or not - it's not strictly in the scope of this change.

File: ecommerce/extensions/payment/views.py (hunk 6 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Updated.

File: ecommerce/extensions/payment/views.py (hunk 7 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py:119]: Updated both. I've made the latter a 400, since CyberSource shouldn't be sending us invalid payments.

File: ecommerce/extensions/payment/views.py (hunk 6 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: This actually shouldn't be a 500, since we're handling CyberSource's response appropriately. Reverted to 200.

Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Scratch that, I was referring to the [PaymentError response](https://github.com/edx/ecommerce/pull/140#discussion-diff-32354343R119).

**Comment by rlucioni** [comment]
@jimabramson I've updated status codes where appropriate. Thanks to you both for the sharp eyes.
